### PR TITLE
plugin vm has tags

### PR DESCRIPTION
**Pull request by abdullahaslam306**

### CODE REVIEW

File: plugins/azure/virtualmachines/vmHasTags.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+var async = require('async');
+
+var helpers = require('../../../helpers/azure/');
+
+module.exports = {
+    title: 'Virtual Machine Has Tags',
+    category: 'Virtual Machines',
+    domain: 'Compute',
+    description: 'Ensures that Azure virtual machine have tags associated.',
```
Comment (alphadev4) [inline plugins/azure/virtualmachines/vmHasTags.js:9]:
```suggestion
    description: 'Ensures that Azure virtual machines have tags associated.',
```

File: plugins/azure/virtualmachines/vmHasTags.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    apis: ['virtualMachines:listAll', 'virtualMachines:get'],
+
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var locations = helpers.locations(settings.govcloud);
+
+        async.each(locations.virtualMachines, function(location, rcb){
+
+            var virtualMachines = helpers.addSource(cache, source, ['virtualMachines', 'listAll', location]);
```
Comment (alphadev4) [inline plugins/azure/virtualmachines/vmHasTags.js:22]:
```suggestion
            var virtualMachines = helpers.addSource(cache, source,
                   ['virtualMachines', 'listAll', location]);
```

Comment (alphadev4) [inline plugins/azure/virtualmachines/vmHasTags.js:22]:
```suggestion
            var virtualMachines = helpers.addSource(cache, source,
                ['virtualMachines', 'listAll', location]);
```

File: plugins/azure/virtualmachines/vmHasTags.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                helpers.addResult(results, 3, 'Unable to query for virtual machines: ' + helpers.addError(virtualMachines), location);
+                return rcb();
+            }
+            if (!virtualMachines.data.length) {
+                helpers.addResult(results, 0, 'No existing virtual machines found', location);
+                return rcb();
+            }
+
+            for (let virtualMachine of virtualMachines.data) { 
+                const virtualMachineData = helpers.addSource(cache, source, ['virtualMachines', 'get', location, virtualMachine.id]);
```
Comment (alphadev4) [inline plugins/azure/virtualmachines/vmHasTags.js:36]:
```suggestion
                const virtualMachineData = helpers.addSource(cache, source,
                    ['virtualMachines', 'get', location, virtualMachine.id]);
```

File: plugins/azure/virtualmachines/vmHasTags.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+                if (!virtualMachineData || !virtualMachineData.data || !virtualMachineData.data.name || virtualMachineData.err) {
+                    helpers.addResult(results, 3, 'unable to query for virtual machine data', location, virtualMachine.id);
+                    continue;
+                } 
+                if (virtualMachineData.data.tags) {
+                    helpers.addResult(results, 0, 'Virtual Machine has tags associated', location, virtualMachine.id);
```
Comment (alphadev4) [inline plugins/azure/virtualmachines/vmHasTags.js:43]:
```suggestion
                    helpers.addResult(results, 0, 'Virtual Machine has tags', location, virtualMachine.id);
```

File: plugins/azure/virtualmachines/vmHasTags.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                } else {
+                    helpers.addResult(results, 2, 'Virtual Machine does not have tags associated', location, virtualMachine.id);
```
Comment (alphadev4) [inline plugins/azure/virtualmachines/vmHasTags.js:45]:
```suggestion
                    helpers.addResult(results, 2, 'Virtual Machine does not have tags', location, virtualMachine.id);
```

File: plugins/azure/virtualmachines/vmHasTags.spec.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                done();
+            });
+        });
+
+        it('should give failing result if VM has does not have tags', function(done) {
+            const cache = createCache([virtualMachines[0]], virtualMachinesData[1]);
+            vmHasTags.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].message).to.include('Virtual Machine does not have tags associated');
```
Comment (alphadev4) [inline plugins/azure/virtualmachines/vmHasTags.spec.js:100]:
```suggestion
                expect(results[0].message).to.include('Virtual Machine does not have tags');
```

File: plugins/azure/virtualmachines/vmHasTags.spec.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                done();
+            });
+        });
+
+        it('should give passing result if VM has tags associated', function(done) {
+            const cache = createCache([virtualMachines[0]], virtualMachinesData[0]);
+            vmHasTags.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('Virtual Machine has tags associated');
```
Comment (alphadev4) [inline plugins/azure/virtualmachines/vmHasTags.spec.js:89]:
```suggestion
                expect(results[0].message).to.include('Virtual Machine has tags');
```

File: plugins/azure/virtualmachines/vmHasTags.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+var helpers = require('../../../helpers/azure/');
+
+module.exports = {
+    title: 'Virtual Machine Has Tags',
+    category: 'Virtual Machines',
+    domain: 'Compute',
+    description: 'Ensures that Azure virtual machines have tags associated.',
+    more_info: 'Tags help you to group resources together that are related to or associated with each other. It is a best practice to tag cloud resources to better organize and gain visibility into their usage.',
+    recommended_action: 'Modify virtual machine and add tags',
```
Comment (alphadev4) [inline plugins/azure/virtualmachines/vmHasTags.js:11]:
```suggestion
    recommended_action: 'Modify affected virtual machine and add tags',
```

File: plugins/azure/virtualmachines/vmHasTags.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+var helpers = require('../../../helpers/azure/');
+
+module.exports = {
+    title: 'Virtual Machine Has Tags',
+    category: 'Virtual Machines',
+    domain: 'Compute',
+    description: 'Ensures that Azure virtual machines have tags associated.',
+    more_info: 'Tags help you to group resources together that are related to or associated with each other. It is a best practice to tag cloud resources to better organize and gain visibility into their usage.',
+    recommended_action: 'Modify affected virtual machine and add tags',
```
Comment (alphadev4) [inline plugins/azure/virtualmachines/vmHasTags.js:11]:
```suggestion
    recommended_action: 'Modify affected virtual machine and add tags.',
```

File: plugins/azure/virtualmachines/vmHasTags.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+                helpers.addResult(results, 3, 'Unable to query for virtual machines: ' + helpers.addError(virtualMachines), location);
+                return rcb();
+            }
+            if (!virtualMachines.data.length) {
+                helpers.addResult(results, 0, 'No existing virtual machines found', location);
+                return rcb();
+            }
+
+            for (let virtualMachine of virtualMachines.data) { 
+                const virtualMachineData = helpers.addSource(cache, source, ['virtualMachines', 'get', location, virtualMachine.id]);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/vmHasTags.js:37]: if not virtualMachine.id check

File: plugins/azure/virtualmachines/vmHasTags.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+                if (!virtualMachineData || !virtualMachineData.data || !virtualMachineData.data.name || virtualMachineData.err) {
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/vmHasTags.js:39]:
```suggestion
                if (!virtualMachineData || !virtualMachineData.data || virtualMachineData.err) {
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**